cmdopt/src/main/scala/pravda/cmdopt/CommandLine.scala

```diff
@@ -0,0 +1,216 @@
+      docref: String = "",
+      verbs: List[Verb[C, _]] = List.empty[Verb[C, _]]
+    ) extends Verb[C, Nothing] {
+
+    def children(xs: Verb[C, _]*): Verb[C, Nothing] = {
+      copy(verbs = verbs ++ xs)
+    }
+
+    def text(msg: String): Cmd[C] = copy(text = msg)
+    def desc(msg: String): Cmd[C] = copy(desc = msg)
+    def action(f: (Any, C) => C): Cmd[C] = copy(action = f)
+    def validate(f: Any => Either[String, Any]): Cmd[C] = copy(validate = f)
+  }
+
+  final case class Opt[C, A: Read](
+      short: Char,
+      name: String,
+      text: String = "",
+      desc: String = "",
+      abbrs: List[String] = List.empty[String],
+      action: (A, C) => C = (_: A, c: C) => c,
+      validate: A => Either[String, Unit] = (_: Any) => Right(()),
+    ) extends Verb[C, A] {
+
+    def read(line: Line, cfg: C): Either[String, (Line, C)] = {
+      val reader = implicitly[Read[A]]
+      reader.read(line) match {
+        case Right((v, newLine)) =>
+          val newCfg = action(v, cfg)
+          Right((newLine, newCfg))
+        case Left(msg) => Left(msg)
+      }
+    }
+
+    def text(msg: String): Opt[C, A] = copy(text = msg)
+    def desc(msg: String): Opt[C,A] = copy(desc = msg)
+    def abbr(name: String): Opt[C, A] = copy(abbrs = abbrs :+ name)
+    def action(f: (A, C) => C): Opt[C, A] = copy(action = f)
+    def validate(f: A => Either[String, Unit]): Opt[C, A] = copy(validate = f)
+  }
+
+  final case class Arg[C, A: Read](
+      name: String,
+      text: String = "",
+      desc: String = "",
+      action: (A, C) => C = (_: A, c: C) => c,
+    ) extends Verb[C, A] {
+
+    def read(line: Line, cfg: C): Either[String, (Line, C)] = {
+      val reader = implicitly[Read[A]]
+      reader.read(line) match {
+        case Right((v, newLine)) =>
+          val newCfg = action(v, cfg)
+          Right((newLine, newCfg))
+        case Left(msg) => Left(msg)
+      }
+    }
+
+    def text(msg: String): Arg[C,A] = copy(text = msg)
+    def desc(msg: String): Arg[C,A] = copy(desc = msg)
+    def action(f: (A, C) => C): Arg[C,A] = copy(action = f)
+  }
+
+  sealed trait Result[+C]
+  final case class Ok[C](config: C) extends Result[C]
+  final case class ParseError(msg: String) extends Result[Nothing]
+  final case class HelpWanted[C](cl: CommandLine[C]) extends Result[Nothing]
+
+  def walk[C](path: List[Cmd[C]], verbs: List[Verb[C, _]]): List[List[Cmd[C]]] = {
+    verbs.collect{ case x: Cmd[_] => x }.flatMap{ cmd =>
+      val newPath = path :+ cmd
+      val list = walk(newPath, cmd.verbs)
+      if (list.isEmpty) {
+        List(newPath)
+      } else {
+        list.map(path ++ _)
+      }
+    }
+  }
+}
+
+trait CommandLine[C] {
+
+  import CommandLine._
+
+  def model: List[Verb[C, _]]
+  def helpOpts = List("-h", "--help", "/?")
+
+  def head(name: String): Head[C] = Head[C](name)
+  def cmd(name: String): Cmd[C] = Cmd[C](name)
+  def opt[A: Read](name: String): Opt[C, A] = Opt[C, A]('\u0000', name)
```

Comment:
`\u0000` looks ugly :) We should use `Option` instead!

Reply:
Fixed!)